migrations/migration_5_share_gob_to_ssz.go

```
@@ -47,23 +70,129 @@ var migration_5_change_share_format_from_gob_to_ssz = Migration{
 			return fmt.Errorf("GetAll: %w", err)
 		}
 
-		if err := opt.Db.SetMany(storagePrefix, len(sets), func(i int) (basedb.Obj, error) {
-			return sets[i], nil
+		sharesGOBTotal = len(sharesGOB)
+		if sharesGOBTotal == 0 {
+			return nil // we won't be creating any SSZ shares
+		}
+
+		if err := opt.Db.SetMany(opstorage.OperatorStoragePrefix, len(sharesSSZEncoded), func(i int) (basedb.Obj, error) {
+			return sharesSSZEncoded[i], nil
 		}); err != nil {
 			return fmt.Errorf("SetMany: %w", err)
 		}
 
-		// TODO - do not complete this migration for now, we'll complete it once
-		// additional sanity-checks are added here
-		//if err := opt.Db.DropPrefix(append(storagePrefix, sharesPrefixGOB...)); err != nil {
-		//	return fmt.Errorf("DropPrefix: %w", err)
-		//}
-		//
-		//// This makes sure migration applies only once.
-		//if err := completed(opt.Db); err != nil {
-		//	return fmt.Errorf("complete transaction: %w", err)
-		//}
+		sharesSSZTotal := 0
+		if err := opt.Db.GetAll(storage.SharesDBPrefix(opstorage.OperatorStoragePrefix), func(i int, obj basedb.Obj) error {
+			shareSSZ := &storage.Share{}
+			err := shareSSZ.Decode(obj.Value)
+			if err != nil {
+				return fmt.Errorf("decode ssz share: %w", err)
+			}
+			sID := shareID(shareSSZ.ValidatorPubKey)
+			shareGOB, ok := sharesGOB[sID]
+			if !ok {
+				// this shouldn't really happen & we should probably return error if it does, but
+				// on stage since we already have some SSV nodes that migrated to SSZ format and
+				// potentially added new validators (new SSZ shares) erroring would prevent migration
+				// from completing, so we don't return error here
+				return nil
+			}
```

I guess this is an unfortunate side-effect of deploying **migration_5** code to **stage** and not completing it later (in this PR)

alternatively, we could "fix that" by merging this PR -> deploying this to all stage nodes -> creating PR that removes this constraint -> merging that new PR & deploying that to prod eventually (but that would mean we will never be able to test this such a change on **stage**)

---

yeah, I think it is better to return an error here (in the next PR as you suggest). Temporarily I would also add some log message, so on stage we could see when this scenario happening

---

> Temporarily I would also add some log message, so on stage we could see when this scenario happening

I don't think there is a need for log message here since it doesn't really tell us anything new (I confirmed already that returning error on **stage** results in exactly what my comment describes) - so logging error here is simply non-actionable, right (hence why even bother logging) ?

---

I see. One question:
For the migrated nodes on Stage, it looks like it wouldn't run into this scenario anyway, because migration 5 was already applied and it seems like if migration was already applied, it [won't be applied again](https://github.com/ssvlabs/ssv/blob/a8258fa62ee8b5f374cf04c2bd77fd7e816cd836/migrations/migrations.go#L82).

---

> For the migrated nodes on Stage, it looks like it wouldn't run into this scenario anyway, because migration 5 was already applied and it seems like if migration was already applied, it [won't be applied again](https://github.com/ssvlabs/ssv/blob/a8258fa62ee8b5f374cf04c2bd77fd7e816cd836/migrations/migrations.go#L82).

your reasoning is correct except that **migration_5** haven't been **completed** for some nodes on stage even though it has been **applied** (because we haven't marked it as **completed** until this PR is merged - https://github.com/ssvlabs/ssv/blob/3e76188c8954fe690b675ee227fd1a507e428656/migrations/migration_5_share_gob_to_ssz.go#L56-L65),

normally you would mark transaction as **completed** in the same PR where it was originally created (not in a separate PR like I'm doing here),

but for this particular **migration_5** we decided to do it this way so we can merge that original PR (it also had other features in it we wanted to have on **stage**) without blocking QA team from testing it for example,

but as I outline in that comment, that results into more "unfortunate inconveniences we have to deal with"

---

understand, thank you for explaining!